File: Programmers/2/87946. 피로도/피로도.java
class Solution {
    private int answer = 0;
    
    public int solution(int k, int[][] dungeons) {
        boolean[] visited = new boolean[dungeons.length];
        
        dfs(k, dungeons, visited, 0);
        
        return answer;
    }
    
    private void dfs(int stamina, int[][] dungeons, boolean[] visited, int count) {
        answer = Math.max(answer, count);
        
        for (int i = 0; i < dungeons.length; i++) {
            int need = dungeons[i][0];
            int cost = dungeons[i][1];
            
            if (visited[i] || stamina < need) continue;
            
            visited[i] = true;
            
            dfs(stamina - cost, dungeons, visited, count + 1);
            
            visited[i] = false;
        }
    }
}

File: Programmers/2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int answer = 0;

void dfs(int stamina, const vector<vector<int>>& dungeon, vector<bool>& visited, int count)
{
    answer = max(answer, count);
    
    for (int i = 0; i < dungeon.size(); i++)
    {
        int need = dungeon[i][0];
        int cost = dungeon[i][1];
        
        if (visited[i] || stamina < need) continue;
        
        visited[i] = true;
        
        dfs(stamina - cost, dungeon, visited, count + 1);
        
        visited[i] = false;
    }
}

int solution(int k, vector<vector<int>> dungeons) {
    vector<bool> visited(dungeons.size(), false);
    
    dfs(k, dungeons, visited, 0);
    
    return answer;
}